src/loader/bgw_message_queue.c

```diff
@@ -0,0 +1,373 @@
+}
+
+
+/* Our own version of shm_mq_wait_for_attach that waits with a timeout so that should our counterparty die before attaching,
+ * we don't end up hanging.*/
+static shm_mq_result
+ts_shm_mq_wait_for_attach(MessageQueue * queue, shm_mq_handle *ack_queue_handle)
+{
+	int			n;
+	PGPROC	   *reader_proc;
+
+
+	for (n = 1; n <= BGW_MQ_NUM_WAITS; n++)
+	{
+		/* The reader is the sender on the ack queue */
+		reader_proc = shm_mq_get_sender(shm_mq_get_queue(ack_queue_handle));
+		if (reader_proc != NULL)
+			return SHM_MQ_SUCCESS;
+		else if (queue_get_reader(queue) == InvalidPid)
+			return SHM_MQ_DETACHED; /* Reader died after we enqueued our
+									 * message */
+#if PG96
+		WaitLatch(MyLatch, WL_LATCH_SET, BGW_MQ_WAIT_INTERVAL);
+#else
+		WaitLatch(MyLatch, WL_LATCH_SET, BGW_MQ_WAIT_INTERVAL, WAIT_EVENT_MQ_INTERNAL);
+#endif
+
+		ResetLatch(MyLatch);
+		CHECK_FOR_INTERRUPTS();
+	}
+	return SHM_MQ_DETACHED;
+
+
+}
+static bool
+enqueue_message_wait_for_ack(MessageQueue * queue, BgwMessage * message, shm_mq_handle *ack_queue_handle)
+{
+	Size		bytes_received = 0;
+	QueueResponseType send_result;
+	bool	   *data = NULL;
+	shm_mq_result mq_res;
+	bool		ack_received = false;
+
+
+	/*
+	 * We don't want the process restarting workers to really distinguish the
+	 * reasons workers might or might not be restarted, and we don't really
+	 * want them to error when workers can't be started, as there are multiple
+	 * valid reasons for that. So we'll simply return false for the ack even
+	 * if we can't attach to the queue etc.
+	 */
+	send_result = queue_add(queue, message);
+	if (send_result != MESSAGE_SENT)
+		return false;
+
+	mq_res = ts_shm_mq_wait_for_attach(queue, ack_queue_handle);
+	if (mq_res != SHM_MQ_SUCCESS)
+		return false;
+
+	/*
+	 * We can use a blocking wait now because if our counterparty detaches we
+	 * will wake
+	 */
+	mq_res = shm_mq_receive(ack_queue_handle, &bytes_received, (void **) &data, false);
+	if (mq_res != SHM_MQ_SUCCESS)
+		return false;
+
+	ack_received = (bytes_received != 0) && *data;
+
+	return ack_received;
+}
+
+
+/*
+ * Write element to queue, wait/error if queue is full
+ * consumes message and deallocates
+ */
+extern bool
+bgw_message_send_and_wait(BgwMessageType message_type, Oid db_oid)
+{
+	shm_mq	   *ack_queue;
+	dsm_segment *seg;
+	shm_mq_handle *ack_queue_handle;
+	BgwMessage *message;
+	bool		ack_received = false;
+
+	message = bgw_message_create(message_type, db_oid);
+
+	seg = dsm_find_mapping(message->ack_dsm_handle);
+	if (seg == NULL)
+		ereport(ERROR, (errmsg("TimescaleDB background worker dynamic shared memory segment not mapped")));
+	ack_queue = shm_mq_create(dsm_segment_address(seg), BGW_ACK_QUEUE_SIZE);
+	shm_mq_set_receiver(ack_queue, MyProc);
+	ack_queue_handle = shm_mq_attach(ack_queue, seg, NULL);
+	if (ack_queue_handle != NULL)
+		ack_received = enqueue_message_wait_for_ack(mq, message, ack_queue_handle);
+	dsm_detach(seg);			/* Queue detach happens in dsm detach callback */
```

Comment:
I think it might be better to call `dsm_detach` as part of a `message_free()` function where you first find the mapping, then detach if not NULL, and  `pfree()` the message.

Reply:
we've checked if it's null above...so not sure we need to deal with that case?